### PR TITLE
NIC: Veth OVS bridge VLAN support

### PR DESCRIPTION
**Pull request by tomponline**
Includes https://github.com/lxc/lxc/pull/3435

Adds openvswitch VLAN support.

### CODE REVIEW

File: src/lxc/confile.c [hunk 1]
```diff
@@ -487,6 +493,69 @@ static int set_config_net_veth_pair(const char *key, const char *value,
...
+				  struct lxc_conf *lxc_conf, void *data)
+{
+	int ret;
+	struct lxc_netdev *netdev = data;
+
+	if (lxc_config_value_empty(value))
+		return clr_config_net_veth_vlan_id(key, lxc_conf, data);
+
+	if (!netdev)
+		return -1;
```
Comment (brauner) [inline src/lxc/confile.c:506]: return ret_errno(EINVAL);

File: src/lxc/confile.c [hunk 2]
```diff
@@ -487,6 +493,69 @@ static int set_config_net_veth_pair(const char *key, const char *value,
...
+	__do_free struct lxc_list *list = NULL;
+	int ret;
+	unsigned short vlan_id;
+	struct lxc_netdev *netdev = data;
+
+	if (lxc_config_value_empty(value))
+		return clr_config_net_veth_vlan_tagged_id(key, lxc_conf, data);
+
+	if (!netdev)
+		return ret_set_errno(-1, EINVAL);
```
Comment (brauner) [inline src/lxc/confile.c:538]: This check should be above the `lxc_config_value_empty()` check as you're passing the `netdev` down to it.

File: src/lxc/confile.c [hunk 3]
```diff
@@ -487,6 +493,69 @@ static int set_config_net_veth_pair(const char *key, const char *value,
...
+	ret = get_u16(&vlan_id, value, 0);
+	if (ret < 0)
+		ret_errno(EINVAL);
+
+	if (vlan_id > BRIDGE_VLAN_ID_MAX)
+		ret_errno(EINVAL);
+
+	list = malloc(sizeof(*list));
+	if (!list)
+		return -1;
```
Comment (brauner) [inline src/lxc/confile.c:549]: return ret_errno(ENOMEM);

File: src/lxc/confile.c [hunk 4]
```diff
@@ -5301,6 +5370,38 @@ static int clr_config_net_veth_pair(const char *key, struct lxc_conf *lxc_conf,
 	return 0;
 }
 
+static int clr_config_net_veth_vlan_id(const char *key, struct lxc_conf *lxc_conf,
+				  void *data)
+{
+	struct lxc_netdev *netdev = data;
+
+	if (!netdev)
+		return -1;
```
Comment (brauner) [inline src/lxc/confile.c:5379]: return ret_errno(EINVAL);

File: src/lxc/confile.c [hunk 5]
```diff
@@ -5301,6 +5370,38 @@ static int clr_config_net_veth_pair(const char *key, struct lxc_conf *lxc_conf,
...
+}
+
+static int clr_config_net_veth_vlan_tagged_id(const char *key,
+				       struct lxc_conf *lxc_conf, void *data)
+{
+	struct lxc_netdev *netdev = data;
+	struct lxc_list *cur, *next;
+
+	if (!netdev)
+		return -1;
```
Comment (brauner) [inline src/lxc/confile.c:5394]: return ret_errno(EINVAL);

File: src/lxc/confile.c [hunk 6]
```diff
@@ -5772,6 +5873,60 @@ static int get_config_net_veth_pair(const char *key, char *retv, int inlen,
...
+	int fulllen = 0;
+	struct lxc_netdev *netdev = data;
+
+	if (!retv)
+		inlen = 0;
+	else
+		memset(retv, 0, inlen);
+
+	if (!netdev)
+		return -1;
```
Comment (brauner) [inline src/lxc/confile.c:5889]: return ret_errno(EINVAL);

and move the two checks above the `if (!retv)` check.

File: src/lxc/network.c [hunk 7]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+	call_cleaner(netlink_close) struct nl_handler *nlh_ptr = &nlh;
+	int err;
+	struct ifinfomsg *ifi;
+	struct rtattr *nest;
+	unsigned short bridge_flags = 0;
+	struct bridge_vlan_info vlan_info;
+
+	err = netlink_open(nlh_ptr, NETLINK_ROUTE);
+	if (err)
+		return ret_errno(-err);
```
Comment (brauner) [inline src/lxc/network.c:266]: This should probably just be:
return err;

File: src/lxc/network.c [hunk 8]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+	}
+
+	return 0;
+}
+
+static int setup_veth_native_bridge_vlan(char *veth1, struct lxc_netdev *netdev)
+{
+	/* Skip setup if no VLAN options are specified. */
+	if (!netdev->priv.veth_attr.vlan_id_set && lxc_list_len(&netdev->priv.veth_attr.vlan_tagged_ids) <= 0)
+		return 0;
```
Comment (brauner) [inline src/lxc/network.c:360]: Move this check after the variable declarations, please.

File: src/lxc/network.c [hunk 9]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+
+	int err, rc, veth1index;
+	char path[STRLITERALLEN("/sys/class/net//bridge/vlan_filtering") + IFNAMSIZ + 1];
+	char buf[5]; // Sufficient size to fit max VLAN ID (4094) and null char.
```
Comment (brauner) [inline src/lxc/network.c:364]: Please use `/* */` only.

File: src/lxc/network.c [hunk 10]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+		if (netdev->priv.veth_attr.vlan_id > BRIDGE_VLAN_NONE) {
+			err = lxc_bridge_vlan_add(veth1index, netdev->priv.veth_attr.vlan_id, false);
+			if (err != 0)
+				return log_error_errno(err, errno, "Failed to add untagged vlan \"%u\" on \"%s\"", netdev->priv.veth_attr.vlan_id, veth1);
+		}
+	}
+
+	/* Configure tagged VLAN settings on bridge port if specified. */
+	err = lxc_bridge_vlan_add_tagged(veth1index, &netdev->priv.veth_attr.vlan_tagged_ids);
+	if (err != 0)
```
Comment (brauner) [inline src/lxc/network.c:419]: `if (err)` everywhere, please. it's the form we use usually.

File: src/lxc/network.c [hunk 11]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+		return log_error_errno(err, errno, "Failed to add tagged vlans on \"%s\"", veth1);
+
+	return 0;
+}
+
+struct ovs_veth_vlan_args {
+	const char *nic;
+	const char *vlan_mode;	// Port VLAN mode.
+	short vlan_id;		// PVID VLAN ID.
+	const char *trunks;	// Comma delimited list of tagged VLAN IDs.
```
Comment (brauner) [inline src/lxc/network.c:429]: `/* */` please

File: src/lxc/network.c [hunk 12]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+};
+
+
+static int lxc_ovs_setup_bridge_vlan_exec(void *data)
+{
+	struct ovs_veth_vlan_args *args = data;
+	const char *vlan_mode = "", *tag = "", *trunks = "";
+
+	vlan_mode = must_concat(NULL, "vlan_mode=", args->vlan_mode, (char *)0);
```
Comment (brauner) [inline src/lxc/network.c:438]:
```c
must_concat(..., ..., ..., (char *)NULL);
```
never `(char *)0`, please.

File: src/lxc/network.c [hunk 13]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+		if (rc < 0 || (size_t)rc >= sizeof(buf))
+			return log_error_errno(-1, EINVAL, "Failed to parse ovs bridge vlan \"%u\"", args->vlan_id);
+
+		tag = must_concat(NULL, "tag=", buf, (char *)0);
+	}
+
+
+	if (strcmp(args->trunks, "") != 0) {
+		trunks = must_concat(NULL, "trunks=", args->trunks, (char *)0);
+	}
```
Comment (brauner) [inline src/lxc/network.c:452]: Remove the `{ }` around single-line if statements, please.

File: src/lxc/network.c [hunk 14]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+
+	/* Detect the combination of vlan_id and trunks specified and convert to ovs-vsctl command. */
+	if (strcmp(tag, "") != 0 && strcmp(trunks, "") != 0)
+		execlp("ovs-vsctl", "ovs-vsctl", "set", "port", args->nic, vlan_mode, tag, trunks, (char *)NULL);
+	else if (strcmp(tag, "") != 0)
+		execlp("ovs-vsctl", "ovs-vsctl", "set", "port", args->nic, vlan_mode, tag, (char *)NULL);
+	else if (strcmp(trunks, "") != 0)
+		execlp("ovs-vsctl", "ovs-vsctl", "set", "port", args->nic, vlan_mode, trunks, (char *)NULL);
+
+	return -1;
```
Comment (brauner) [inline src/lxc/network.c:462]: You should surface the errno from exec, i.e. add another else which returns `-EINVAL` and replace the `return -1` with `return -errno`;

File: src/lxc/network.c [hunk 15]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+	/* Configure untagged VLAN settings on bridge port if specified. */
+	if (netdev->priv.veth_attr.vlan_id_set) {
+		if (netdev->priv.veth_attr.vlan_id == BRIDGE_VLAN_NONE && taggedLength <= 0)
+			return log_error_errno(-1, EINVAL, "Cannot use vlan.id=none with openvswitch bridges when not using vlan.tagged.id");
+
+		// Configure the untagged 'native' membership settings of the port if VLAN ID specified.
+		// Also set the vlan_mode=access, which will drop any tagged frames.
+		// Order is important here, as vlan_mode is set to "access", assuming that vlan.tagged.id is not
+		// used. If vlan.tagged.id is specified, then we expect it to also change the vlan_mode as needed.
+		if (netdev->priv.veth_attr.vlan_id > BRIDGE_VLAN_NONE) {
```
Comment (brauner) [inline src/lxc/network.c:487]: `/* */` please

File: src/lxc/network.c [hunk 16]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+			args.vlan_mode = "access";
+			args.vlan_id = netdev->priv.veth_attr.vlan_id;
+		}
+	}
+
+	if (taggedLength > 0) {
+		args.vlan_mode = "trunk"; // Default to only allowing tagged frames (drop untagged frames).
```
Comment (brauner) [inline src/lxc/network.c:494]: `/* */`

File: src/lxc/network.c [hunk 17]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+
+		if (netdev->priv.veth_attr.vlan_id > BRIDGE_VLAN_NONE) {
+			// If untagged vlan mode isn't "none" then allow untagged frames for port's 'native' VLAN.
```
Comment (brauner) [inline src/lxc/network.c:497]: `/* */`

File: src/lxc/network.c [hunk 18]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+			args.vlan_mode  = "native-untagged";
+		}
+
+		struct lxc_list *iterator;
+		lxc_list_for_each(iterator, &netdev->priv.veth_attr.vlan_tagged_ids) {
+			char buf[5]; // Sufficient size to fit max VLAN ID (4094) null char.
+			unsigned short vlan_id = PTR_TO_USHORT(iterator->elem);
+			int rc = snprintf(buf, sizeof(buf), "%u", vlan_id);
```
Comment (brauner) [inline src/lxc/network.c:505]:
```c
			unsigned short vlan_id = PTR_TO_USHORT(iterator->elem);
                        char buf[5]; /* Sufficient size to fit max VLAN ID (4094) null char. */
			int rc;

                        rc = snprintf(buf, sizeof(buf), "%u", vlan_id);
```

File: src/lxc/network.c [hunk 19]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+			if (rc < 0 || (size_t)rc >= sizeof(buf))
+				return log_error_errno(-1, EINVAL, "Failed to parse tagged vlan \"%u\" for interface \"%s\"", vlan_id, veth1);
+
+			args.trunks = must_concat(NULL, args.trunks, buf, ",", (char *)0);
+		}
+	}
+
+	if (strcmp(args.vlan_mode, "") != 0) {
+		char cmd_output[PATH_MAX];
+		int ret = run_command(cmd_output, sizeof(cmd_output), lxc_ovs_setup_bridge_vlan_exec, (void *)&args);
```
Comment (brauner) [inline src/lxc/network.c:515]:
```c
                int ret;
		char cmd_output[PATH_MAX];

		ret = run_command(cmd_output, sizeof(cmd_output), lxc_ovs_setup_bridge_vlan_exec, (void *)&args);
```

File: src/lxc/network.c [hunk 20]
```diff
@@ -245,13 +245,292 @@ static int lxc_is_ip_forwarding_enabled(const char *ifname, int family)
...
+
 static int instantiate_veth(struct lxc_handler *handler, struct lxc_netdev *netdev)
 {
 	int err;
 	unsigned int mtu = 1500;
 	char *veth1, *veth2;
 	char veth1buf[IFNAMSIZ], veth2buf[IFNAMSIZ];
 
+	err = validate_veth(netdev);
+	if (err != 0)
```
Comment (brauner) [inline src/lxc/network.c:531]: if (err)

File: src/lxc/network.c [hunk 21]
```diff
@@ -324,14 +603,33 @@ static int instantiate_veth(struct lxc_handler *handler, struct lxc_netdev *netd
...
-			SYSERROR("Failed to attach \"%s\" to bridge \"%s\"",
-			         veth1, netdev->link);
+			SYSERROR("Failed to attach \"%s\" to bridge \"%s\"", veth1, netdev->link);
 			goto out_delete;
 		}
 		INFO("Attached \"%s\" to bridge \"%s\"", veth1, netdev->link);
+
+		if (is_ovs_bridge(netdev->link)) {
+			err = setup_veth_ovs_bridge_vlan(veth1, netdev);
+			if (err != 0) {
```
Comment (brauner) [inline src/lxc/network.c:621]: if (err)

File: src/lxc/network.c [hunk 22]
```diff
@@ -324,14 +603,33 @@ static int instantiate_veth(struct lxc_handler *handler, struct lxc_netdev *netd
...
+				SYSERROR("Failed to setup openvswitch bridge vlan on \"%s\"", veth1);
+				lxc_ovs_delete_port(netdev->link, veth1);
+				goto out_delete;
+			}
+		} else {
+			err = setup_veth_native_bridge_vlan(veth1, netdev);
+			if (err != 0) {
```
Comment (brauner) [inline src/lxc/network.c:628]: if (err)

**Comment by tomponline** [comment]
jenkins: test this please

**Pull request merged**